plugins/rest/azure.go

```diff
@@ -0,0 +1,139 @@
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"time"
+)
+
+var (
+	azureIMDSEndpoint = "http://169.254.169.254/metadata/identity/oauth2/token"
+	defaultApiVersion = "2018-02-01"
+	timeout           = time.Duration(5) * time.Second
+)
+
+// azureManagedIdentitiesToken holds a token for managed identities for Azure resources
+type azureManagedIdentitiesToken struct {
+	AccessToken  string `json:"access_token"`
+	ExpiresIn    string `json:"expires_in"`
+	ExpiresOn    string `json:"expires_on"`
+	NotBefore    string `json:"not_before"`
+	Resource     string `json:"resource"`
+	TokenType    string `json:"token_type"`
+}
+
+// azureManagedIdentitiesError represents an error fetching an azureManagedIdentitiesToken
+type azureManagedIdentitiesError struct {
+	err         string `json:"error"`
+	description string `json:"error_description"`
+	endpoint    string
+	statusCode  int
+}
+
+func (e *azureManagedIdentitiesError) Error() string {
+	return fmt.Sprintf("%v %s retrieving azure token from %s: %s", e.statusCode, e.err, e.endpoint, e.description)
+}
+
+// azureManagedIdentitiesAuthPlugin uses an azureManagedIdentitiesToken.AccessToken for bearer authorization
+type azureManagedIdentitiesAuthPlugin struct {
+	Endpoint   string `json:"endpoint"`
+	APIVersion string `json:"api_version"`
+	Resource   string `json:"resource"`
+	ObjectID   string `json:"object_id"`
+	ClientID   string `json:"client_id"`
+	MiResID    string `json:"mi_res_id"`
+}
+
+func (ap *azureManagedIdentitiesAuthPlugin) NewClient(c Config) (*http.Client, error) {
+	if ap.Resource == "" {
+		return nil, errors.New("resource URI is required when the azure managed identities plugin is enabled")
+	}
+
+	if ap.APIVersion == "" {
+		ap.APIVersion = defaultApiVersion
+	}
+
+	if ap.Endpoint == "" {
+		ap.Endpoint = azureIMDSEndpoint
+	}
+
+	t, err := DefaultTLSConfig(c)
+	if err != nil {
+		return nil, err
+	}
+
+	return DefaultRoundTripperClient(t, *c.ResponseHeaderTimeoutSeconds), nil
+}
+
+func (ap *azureManagedIdentitiesAuthPlugin) Prepare(req *http.Request) error {
+	token, err := azureManagedIdentitiesTokenRequest(
+		ap.Endpoint, ap.APIVersion, ap.Resource,
+		ap.ObjectID, ap.ClientID, ap.MiResID,
+	)
+	if err != nil {
+		return err
+	}
+
+	req.Header.Add("Authorization", fmt.Sprintf("Bearer %v", token.AccessToken))
+	return nil
+}
+
+// azureManagedIdentitiesTokenRequest fetches an azureManagedIdentitiesToken
+func azureManagedIdentitiesTokenRequest(
+	endpoint, apiVersion, resource, objectId, clientId, miResId string,
+) (azureManagedIdentitiesToken, error) {
+	e := fmt.Sprintf("%s?api-version=%s&resource=%s", endpoint, apiVersion, resource)
+
+	if objectId != "" {
+		e += fmt.Sprintf("&object_id=%s", objectId)
+	}
+
+	if clientId != "" {
+		e += fmt.Sprintf("&client_id=%s", clientId)
+	}
+
+	if miResId != "" {
+		e += fmt.Sprintf("&mi_res_id=%s", miResId)
+	}
+
+	request, err := http.NewRequest("GET", e, nil)
+	if err != nil {
+		return azureManagedIdentitiesToken{}, err
```

Comment:
Why return an empty struct here rather than nil?

Reply:
I don't think we can return `nil` for a straight `struct` type return. I've used this approach in the past, but I agree it's quite ugly and there's a few other options:

* We can declare an empty `var token azureManagedIdentitiesToken` at the top of the method and return that whenever we have a non-nil error. 
* We can change the return type to `*azureManagedIdentitiesToken`. For error cases we would return `nil`, and for success cases we would return a reference. 

I like the first option, what do you think?

Reply:
I've taken a look around the repo and I think the first option is used. I'm going to go ahead and implement it, but leave this conversation unresolved. Feel free to resolve it or continue the discussion.

Reply:
Resolving for now.